crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -76,139 +78,106 @@ async fn sys_validation_workflow_inner(
-                let op_type = op.get_type();
-                let action = op.action();
-
-                let dependency = get_dependency(op_type, &action);
-                let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
-
-                let r = validate_op(&op, &dna_def, &cascade, Some(incoming_dht_ops_sender)).await;
-                r.map(|o| (op_hash, o, dependency))
-            }
-        }
-    });
-
-    // Create a stream of concurrent validation futures.
-    // This will run NUM_CONCURRENT_OPS validation futures concurrently and
-    // return up to NUM_CONCURRENT_OPS * 100 results.
-    use futures::stream::StreamExt;
-    let mut iter = futures::stream::iter(iter)
-        .buffer_unordered(NUM_CONCURRENT_OPS)
-        .ready_chunks(NUM_CONCURRENT_OPS * 100);
-
-    // Spawn a task to actually drive the stream.
-    // This allows the stream to make progress in the background while
-    // we are committing previous results to the database.
-    let (tx, rx) = tokio::sync::mpsc::channel(NUM_CONCURRENT_OPS * 100);
-    let jh = tokio::spawn(async move {
-        // Send the result to task that will commit to the database.
-        while let Some(op) = iter.next().await {
-            if tx
-                .send_timeout(op, std::time::Duration::from_secs(10))
-                .await
-                .is_err()
-            {
-                tracing::warn!("sys validation task has failed to send ops. This is not a problem if the conductor is shutting down");
-                break;
+    validate_ops_batch(
+        sorted_ops,
+        start,
+        {
+            let space = space.clone();
+            move |so| {
+                // Create an incoming ops sender for any dependencies we find
+                // that we are meant to be holding but aren't.
+                // If we are not holding them they will be added to our incoming ops.
+                let incoming_dht_ops_sender =
+                    IncomingDhtOpSender::new(space.clone(), sys_validation_trigger.clone());
+                let workspace = workspace.clone();
+                let cascade = cascade.clone();
+                async move {
+                    let (op, op_hash) = so.into_inner();
+                    let op_type = op.get_type();
+                    let action = op.action();
+
+                    let dependency = get_dependency(op_type, &action);
+                    let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
+
+                    let r =
+                        validate_op(&op, &dna_def, &cascade, Some(incoming_dht_ops_sender)).await;
+                    r.map(|o| (op_hash, o, dependency))
+                }
+                .boxed()
             }
-        }
-    });
-
-    // Create a stream that will chunk up to NUM_CONCURRENT_OPS * 100 ready results.
-    let mut iter =
-        tokio_stream::wrappers::ReceiverStream::new(rx).ready_chunks(NUM_CONCURRENT_OPS * 100);
```

Comment:
This piece of code has been deleted, along with the Tokio task above that was feeding it. As far as I can tell this won't do anything useful under a normal load. In the worst case it creates very long database transaction by taking (NUM_CONCURRENT_OPS * 100) ^ 2 = 25_000_000 ops and trying to write them to the database. 

One layer of batch+buffer makes sense, two layers is not adding more performance in reality because it won't actually chunk here unless multiple sends to the input channel happen in the same Tokio cycle. It will just forward whatever it has on each poll. Even in the first layer it's most common to see a single chunk get forwarded and that may not even be full. 

The only time where you'll see the buffering be useful I believe is if it takes some time get a database transaction because Holochain is busy. If that's the case it's better to interleave batches with whatever else is going on and 5k is plenty, no need to try to write to 25m rows at a time.

Reply:
For sure and it was a really difficult one to wrap my head around.

Reply:
Tell me about it! I spent a good chunk of time staring at this and it wasn't until I had tests running against it that I started to feel like I understood it

Reply:
Yep, I needed logs too